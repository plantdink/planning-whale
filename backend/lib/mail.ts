import { createTransport, getTestMessageUrl } from "nodemailer";

const transport = createTransport({
  host: String(process.env.MAIL_HOST),
  port: Number(process.env.MAIL_PORT),
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function emailTemplate(text: string) {
  return `
  <div className="email" style="
    border: 1px solid black;
    padding: 20px;
    font-family: sans-serif;
    line-height: 2;
    font-size: 20px;
    ">
    <h2>Password Reset</h2>
    <p>${text}</p>
    <p>Admin team</p>
    </div>
  `;
}

export interface Envelope {
  from: string;
  to?: string[] | null;
}

export interface MailResponse {
  accepted?: string[] | null;
  rejected?: null[] | null;
  envelopeTime: number;
  messageTime: number;
  messageSize: number;
  response: string;
  envelope: Envelope;
  messageId: string;
}

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const info = (await transport.sendMail({
    to,
    from: "Admin@agentfieldmanual.com",
    subject: "Your password reset token",
    html: emailTemplate(`Here is your password reset token
    <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click here to reset your admin account password</a>
    `),
  })) as MailResponse;
  if (process.env.MAIL_USER.includes("ethereal.email")) {
    console.log(`Message Sent. Preview it at ${getTestMessageUrl(info)}`);
  }
}
