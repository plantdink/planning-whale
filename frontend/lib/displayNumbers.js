export function displayPercentage(number = 0) {
  const displayNumber = number / 10;
  return displayNumber;
}

export function hsDisplay(number = 0) {
  const displayNumber = number / 100;
  return displayNumber;
}
