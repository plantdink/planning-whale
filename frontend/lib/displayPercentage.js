export default function displayPercentage(number = 0) {
  const displayNumber = number / 10;
  return displayNumber;
}
