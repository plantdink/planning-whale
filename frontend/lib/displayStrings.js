export function lowerCaseFirstLetter(string) {
  const lowerCaseString = string.toLowerCase();
  const titleString = lowerCaseString.replace(
    /(^\w{1})|(\s+\w{1})/g,
    (letter) => letter.toUpperCase()
  );
  const noWhiteSpaceString = titleString.replaceAll(/\s/g, '');
  const finalLink =
    noWhiteSpaceString.charAt(0).toLowerCase() + noWhiteSpaceString.slice(1);
  return finalLink;
}

export function titleCase(string) {
  return string
    .toLowerCase()
    .split(' ')
    .map(function (word) {
      return word.charAt(0).toUpperCase() + word.slice(1);
    })
    .join(' ');
}
