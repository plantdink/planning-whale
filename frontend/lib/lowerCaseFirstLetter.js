export default function lowerCaseFirstLetter(string) {
  const lowerCaseString = string.toLowerCase();
  const titleString = lowerCaseString.replace(
    /(^\w{1})|(\s+\w{1})/g,
    (letter) => letter.toUpperCase()
  );
  const noWhiteSpaceString = titleString.replaceAll(/\s/g, '');
  const finalLink =
    noWhiteSpaceString.charAt(0).toLowerCase() + noWhiteSpaceString.slice(1);
  return finalLink;
}
