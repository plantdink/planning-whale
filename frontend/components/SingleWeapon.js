import React from 'react';
import {
  Chart as ChartJS,
  RadialLinearScale,
  PointElement,
  LineElement,
  Filler,
  Tooltip,
  Legend,
} from 'chart.js';
import { Radar } from 'react-chartjs-2';
import {
  displayPercentage,
  hsDisplay,
  millisecondsToSeconds,
  humanReadableNumber,
} from '../lib/displayNumbers';
import HeadSEO from './HeadSEO';
import { titleCase } from '../lib/displayStrings';
import { deAbbreviate } from '../lib/abbreviationsDictionary';

export default function SingleWeapon({ weapon, avgWeapon }) {
  ChartJS.register(
    RadialLinearScale,
    PointElement,
    LineElement,
    Filler,
    Tooltip,
    Legend
  );

  // this sets the default size for the whole chart & tooltips, doesn't affect the individual point labels
  ChartJS.defaults.font.size = 14;

  const options = {
    responsive: true,
    scales: {
      r: {
        ticks: {
          display: false,
        },
      },
    },
    plugins: {
      legend: {
        position: 'top',
      },
      tooltip: {
        callbacks: {
          label(context) {
            const weaponDataSet = [
              `${displayPercentage(weapon.accuracy)}%`,
              `${displayPercentage(weapon.stability)}%`,
              weapon.magazineSize,
              weapon.rpm,
              `${weapon.optimalRange}m`,
              `${weapon.maxRange}m`,
              weapon.modSlots,
              `${millisecondsToSeconds(weapon.reloadSpeed)} sec`,
              `${millisecondsToSeconds(weapon.reloadSpeedFromEmpty)} sec`,
              `${humanReadableNumber(weapon.damageWt5) || 0}`,
              `${humanReadableNumber(weapon.damageLevel40)}`,
            ];

            const AverageWeaponDataSet = [
              `${displayPercentage(avgWeapon.accuracy)}%`,
              `${displayPercentage(avgWeapon.stability)}%`,
              avgWeapon.magazineSize,
              avgWeapon.rpm,
              `${avgWeapon.optimalRange}m`,
              `${avgWeapon.maxRange}m`,
              avgWeapon.modSlots,
              `${millisecondsToSeconds(avgWeapon.reloadSpeed)} sec`,
              `${millisecondsToSeconds(avgWeapon.reloadSpeedFromEmpty)} sec`,
              `${humanReadableNumber(avgWeapon.damageWT5)}`,
              `${humanReadableNumber(avgWeapon.damageLevel40)}`,
            ];

            let labelValues = [];

            if (context.dataset.label === weapon.model)
              labelValues = weaponDataSet;
            if (
              context.dataset.label === `Average ${titleCase(avgWeapon.class)}`
            )
              labelValues = AverageWeaponDataSet;
            const label = labelValues[context.dataIndex];
            return label;
          },
          beforeTitle(context) {
            return context[0].dataset.label;
          },
        },
      },
    },
  };

  const chartLabels = [
    'Accuracy',
    'Stability',
    'Magazine Size',
    'RPM',
    'Optimal Range',
    'Max Range',
    'Mod Slots',
    'Reload Speed',
    'Reload from empty',
    'Damage (World Tier 5)',
    'Damage (Level 40)',
  ];

  const weaponDatasetComputed = [
    (weapon.accuracy / avgWeapon.accuracy) * 100,
    (weapon.stability / avgWeapon.stability) * 100,
    (weapon.magazineSize / avgWeapon.magazineSize) * 100,
    (weapon.rpm / avgWeapon.rpm) * 100,
    (weapon.optimalRange / avgWeapon.optimalRange) * 100,
    (weapon.maxRange / avgWeapon.maxRange) * 100,
    (weapon.modSlots / avgWeapon.modSlots) * 100,
    (weapon.reloadSpeed / avgWeapon.reloadSpeed) * 100,
    (weapon.reloadSpeedFromEmpty / avgWeapon.reloadSpeedFromEmpty) * 100,
    (weapon.damageWt5 / avgWeapon.damageWT5) * 100,
    (weapon.damageLevel40 / avgWeapon.damageLevel40) * 100,
  ];

  const averageWeaponDatasetComputed = [
    (avgWeapon.accuracy / avgWeapon.accuracy) * 100,
    (avgWeapon.stability / avgWeapon.stability) * 100,
    (avgWeapon.magazineSize / avgWeapon.magazineSize) * 100,
    (avgWeapon.rpm / avgWeapon.rpm) * 100,
    (avgWeapon.optimalRange / avgWeapon.optimalRange) * 100,
    (avgWeapon.maxRange / avgWeapon.maxRange) * 100,
    (avgWeapon.modSlots / avgWeapon.modSlots) * 100,
    (avgWeapon.reloadSpeed / avgWeapon.reloadSpeed) * 100,
    (avgWeapon.reloadSpeedFromEmpty / avgWeapon.reloadSpeedFromEmpty) * 100,
    (avgWeapon.damageWT5 / avgWeapon.damageWT5) * 100,
    (avgWeapon.damageLevel40 / avgWeapon.damageLevel40) * 100,
  ];

  // change the background colour of the chart dataset fill depending on weapon type
  let backgroundColorComputed = 'rgba(255, 109, 16, 0.6)';
  if (weapon.isExotic === 'YES')
    backgroundColorComputed = 'rgba(255, 111, 54, 0.6)';
  if (weapon.isNamed === 'YES')
    backgroundColorComputed = 'rgba(234, 162, 19, 0.6';

  // change the colour of the chart dataset border line depending on weapon type
  let borderColorComputed = 'rgba(255, 109, 16, 1)';
  if (weapon.isExotic === 'YES') borderColorComputed = 'rgba(255, 111, 54, 1)';
  if (weapon.isNamed === 'YES') borderColorComputed = 'rgba(234, 162, 19, 1)';

  const chartData = {
    labels: chartLabels,
    datasets: [
      {
        label: `${weapon.model}`,
        data: weaponDatasetComputed,
        backgroundColor: backgroundColorComputed,
        borderColor: borderColorComputed,
        borderWidth: 2,
      },
      {
        label: `Average ${titleCase(avgWeapon.class)}`,
        data: averageWeaponDatasetComputed,
        backgroundColor: 'rgba(58, 58, 58, 0.3)',
        borderColor: 'rgba(58, 58, 58, 0.8)',
        borderWidth: 1,
      },
    ],
  };

  return (
    <>
      <HeadSEO seoTag={weapon.model} />
      <div className="single-weapon__content">
        <div className="single-weapon__details">
          <div className="single-weapon__title-bar">
            <h1 className="single-weapon__heading">{weapon.model}</h1>
          </div>
          {(weapon.flavourText !== '' || null) && (
            <blockquote>{weapon.flavourText}</blockquote>
          )}

          <p>
            {titleCase(weapon.class)} {weapon.family}
          </p>
          <div className="single-weapon__title-bar">
            <h3 className="single-weapon__subheading">Base Damage</h3>
          </div>
          <p>{humanReadableNumber(weapon.damageLevel40)} (Level 40)</p>
          <p>{humanReadableNumber(weapon.damageWt5) || 'N/A'} (World Tier 5)</p>
          <div className="single-weapon__title-bar">
            <h3 className="single-weapon__subheading">Specifications</h3>
          </div>
          {weapon.maxBonusValue > 1 && (
            <p>
              {weapon.maxBonusValue}% {deAbbreviate(weapon.weaponBonusType)}
            </p>
          )}
          <p>{hsDisplay(weapon.headshotMultiplier)} x Headshot Multiplier</p>
          <p>{weapon.rpm} rounds per minute rate of fire</p>
          <p>{weapon.magazineSize} round magazine</p>
          {weapon.modSlots === 0 && <p>No adjustable modification slots</p>}
          {weapon.modSlots > 0 && <p>{weapon.modSlots} modification slots</p>}
          <p>{displayPercentage(weapon.accuracy)}% accuracy</p>
          <p>{displayPercentage(weapon.stability)}% stability</p>
          <p>{weapon.optimalRange}m optimal range</p>
          <p>{weapon.maxRange}m max range</p>
          <p>
            {millisecondsToSeconds(weapon.reloadSpeed)} seconds reload speed
          </p>
          <p>
            {millisecondsToSeconds(weapon.reloadSpeedFromEmpty)} seconds reload
            speed (empty magazine)
          </p>
          {(!weapon.accuracy || !weapon.stability) && (
            <>
              <small>Note: 0 indicates no value available</small>
            </>
          )}
        </div>
        <div className="single-weapon__details">
          <div className="single-weapon__title-bar">
            <h3 className="single-weapon__subheading">Stats Comparison</h3>
          </div>

          <Radar data={chartData} options={options} />
        </div>
      </div>
    </>
  );
}
