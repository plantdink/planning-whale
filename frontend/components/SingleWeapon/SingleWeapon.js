import React from 'react';
import { gql } from '@apollo/client';
import {
  Chart as ChartJS,
  RadialLinearScale,
  PointElement,
  LineElement,
  Filler,
  Tooltip,
  Legend,
} from 'chart.js';
import { Radar } from 'react-chartjs-2';
import styled from 'styled-components';
import {
  displayPercentage,
  hsDisplay,
  millisecondsToSeconds,
  humanReadableNumber,
} from '../../lib/displayNumbers';
import HeadSEO from '../HeadSEO';
import { titleCase } from '../../lib/displayStrings';
import { deAbbreviate } from '../../lib/abbreviationsDictionary';

export const SingleWeaponStyles = styled.div`
  position: relative;
  display: flex;
  flex-flow: column wrap;
  max-width: var(--maxWidth);
  background: white;
  border: 1px solid var(--offWhite);
  box-shadow: var(--bs);
  padding: 2rem;
  margin-bottom: 3rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;

  .single-weapon__content {
    display: grid;
    grid-auto-columns: 1fr 3fr;
    grid-auto-flow: column;
    max-width: var(--maxWidth);
    justify-content: center;
    align-items: top;
    gap: 2rem;
    background: white;
    img {
      width: 100%;
      object-fit: contain;
    }
  }

  .single-weapon__details {
  }

  .single-weapon__title-bar {
    transform: skew(7deg);
    margin-bottom: 1rem;
  }

  .single-weapon__heading {
    margin: 0 1rem;
    margin-top: -3rem;
    background: var(--orange);
    display: inline;
    line-height: 1.3;
    font-size: clamp(2rem, 5vw, 4rem);
    color: var(--text);
    /* text-shadow: 2px 2px 0 rgba(0, 0, 0, 0.1); */
    padding: 0 1rem;
  }

  .single-weapon__subheading {
    margin: 0 1rem;
    /* margin-top: 0rem; */
    background: var(--orange);
    display: inline;
    line-height: 2;
    font-size: clamp(1rem, 3vw, 2rem);
    font-weight: 600;
    color: var(--text);
    /* text-shadow: 2px 2px 0 rgba(0, 0, 0, 0.1); */
    padding: 0 1rem;
  }

  p {
    color: var(--text);
    margin: 0.75rem 0;
    font-weight: 400;
    padding: 0 2.4rem;
    font-size: 1.6rem;
  }

  blockquote {
    color: var(--text);
    margin: 0.85rem 0;
    padding: 0 2.4rem;
    font-weight: 300;
    font-size: 1.6rem;
  }

  small {
    color: var(--text);
    font-size: 1.2rem;
    padding: 0 2.4rem;
  }

  .named-weapon {
    background: var(--namedGold);
  }

  .exotic-weapon {
    background: var(--exoticOrange);
  }

  @media (max-width: 35em) {
    padding: 0;
    .single-weapon__content {
      padding-top: 1rem;
      padding-bottom: 1rem;
      display: flex;
      flex-direction: column;
    }
  }
`;

export const SINGLE_WEAPON_QUERY = gql`
  query SINGLE_WEAPON_QUERY($id: ID!) {
    allWeapons(where: { id: $id }) {
      id
      class
      model
      family
      magazineSize
      rpm
      modSlots
      reloadSpeed
      reloadSpeedFromEmpty
      accuracy
      stability
      optimalRange
      maxRange
      headshotMultiplier
      weaponBonusType
      maxBonusValue
      damageLevel40
      damageWt5
      flavourText
      isNamed
      isExotic
      weaponTalent {
        id
        name
        image {
          image {
            publicUrlTransformed
          }
        }
      }
      averageWeapon {
        class
        magazineSize
        rpm
        modSlots
        reloadSpeed
        reloadSpeedFromEmpty
        accuracy
        stability
        optimalRange
        maxRange
        headshotMultiplier
        damageLevel40
        damageWT5
      }
    }
  }
`;

export default function SingleWeapon({ weapon, avgWeapon }) {
  ChartJS.register(
    RadialLinearScale,
    PointElement,
    LineElement,
    Filler,
    Tooltip,
    Legend
  );

  // this sets the default size for the whole chart & tooltips, doesn't affect the individual point labels
  ChartJS.defaults.font.size = 14;

  const options = {
    responsive: true,
    scales: {
      r: {
        ticks: {
          display: false,
        },
      },
    },
    plugins: {
      legend: {
        position: 'top',
      },
      tooltip: {
        callbacks: {
          label(context) {
            const weaponDataSet = [
              `${displayPercentage(weapon.accuracy)}%`,
              `${displayPercentage(weapon.stability)}%`,
              weapon.magazineSize,
              weapon.rpm,
              `${weapon.optimalRange}m`,
              `${weapon.maxRange}m`,
              weapon.modSlots,
              `${millisecondsToSeconds(weapon.reloadSpeed)} sec`,
              `${millisecondsToSeconds(weapon.reloadSpeedFromEmpty)} sec`,
              `${humanReadableNumber(weapon.damageWt5) || 0}`,
              `${humanReadableNumber(weapon.damageLevel40)}`,
            ];

            const AverageWeaponDataSet = [
              `${displayPercentage(avgWeapon.accuracy)}%`,
              `${displayPercentage(avgWeapon.stability)}%`,
              avgWeapon.magazineSize,
              avgWeapon.rpm,
              `${avgWeapon.optimalRange}m`,
              `${avgWeapon.maxRange}m`,
              avgWeapon.modSlots,
              `${millisecondsToSeconds(avgWeapon.reloadSpeed)} sec`,
              `${millisecondsToSeconds(avgWeapon.reloadSpeedFromEmpty)} sec`,
              `${humanReadableNumber(avgWeapon.damageWT5)}`,
              `${humanReadableNumber(avgWeapon.damageLevel40)}`,
            ];

            let labelValues = [];

            if (context.dataset.label === weapon.model)
              labelValues = weaponDataSet;
            if (
              context.dataset.label === `Average ${titleCase(avgWeapon.class)}`
            )
              labelValues = AverageWeaponDataSet;
            const label = labelValues[context.dataIndex];
            return label;
          },
          beforeTitle(context) {
            return context[0].dataset.label;
          },
        },
      },
    },
  };

  const chartLabels = [
    'Accuracy',
    'Stability',
    'Magazine Size',
    'RPM',
    'Optimal Range',
    'Max Range',
    'Mod Slots',
    'Reload Speed',
    'Reload from empty',
    'Damage (World Tier 5)',
    'Damage (Level 40)',
  ];

  const weaponDatasetComputed = [
    (weapon.accuracy / avgWeapon.accuracy) * 100,
    (weapon.stability / avgWeapon.stability) * 100,
    (weapon.magazineSize / avgWeapon.magazineSize) * 100,
    (weapon.rpm / avgWeapon.rpm) * 100,
    (weapon.optimalRange / avgWeapon.optimalRange) * 100,
    (weapon.maxRange / avgWeapon.maxRange) * 100,
    (weapon.modSlots / avgWeapon.modSlots) * 100,
    (weapon.reloadSpeed / avgWeapon.reloadSpeed) * 100,
    (weapon.reloadSpeedFromEmpty / avgWeapon.reloadSpeedFromEmpty) * 100,
    (weapon.damageWt5 / avgWeapon.damageWT5) * 100,
    (weapon.damageLevel40 / avgWeapon.damageLevel40) * 100,
  ];

  const averageWeaponDatasetComputed = [
    (avgWeapon.accuracy / avgWeapon.accuracy) * 100,
    (avgWeapon.stability / avgWeapon.stability) * 100,
    (avgWeapon.magazineSize / avgWeapon.magazineSize) * 100,
    (avgWeapon.rpm / avgWeapon.rpm) * 100,
    (avgWeapon.optimalRange / avgWeapon.optimalRange) * 100,
    (avgWeapon.maxRange / avgWeapon.maxRange) * 100,
    (avgWeapon.modSlots / avgWeapon.modSlots) * 100,
    (avgWeapon.reloadSpeed / avgWeapon.reloadSpeed) * 100,
    (avgWeapon.reloadSpeedFromEmpty / avgWeapon.reloadSpeedFromEmpty) * 100,
    (avgWeapon.damageWT5 / avgWeapon.damageWT5) * 100,
    (avgWeapon.damageLevel40 / avgWeapon.damageLevel40) * 100,
  ];

  // change the background colour of the chart dataset fill depending on weapon type
  let backgroundColorComputed = 'rgba(255, 109, 16, 0.6)';
  if (weapon.isExotic === 'YES')
    backgroundColorComputed = 'rgba(255, 111, 54, 0.6)';
  if (weapon.isNamed === 'YES')
    backgroundColorComputed = 'rgba(234, 162, 19, 0.6';

  // change the colour of the chart dataset border line depending on weapon type
  let borderColorComputed = 'rgba(255, 109, 16, 1)';
  if (weapon.isExotic === 'YES') borderColorComputed = 'rgba(255, 111, 54, 1)';
  if (weapon.isNamed === 'YES') borderColorComputed = 'rgba(234, 162, 19, 1)';

  const chartData = {
    labels: chartLabels,
    datasets: [
      {
        label: `${weapon.model}`,
        data: weaponDatasetComputed,
        backgroundColor: backgroundColorComputed,
        borderColor: borderColorComputed,
        borderWidth: 2,
      },
      {
        label: `Average ${titleCase(avgWeapon.class)}`,
        data: averageWeaponDatasetComputed,
        backgroundColor: 'rgba(58, 58, 58, 0.3)',
        borderColor: 'rgba(58, 58, 58, 0.8)',
        borderWidth: 1,
      },
    ],
  };

  return (
    <>
      <HeadSEO seoTag={weapon.model} />
      <div className="single-weapon__content">
        <div className="single-weapon__details">
          <div className="single-weapon__title-bar">
            <h1 className="single-weapon__heading">{weapon.model}</h1>
          </div>
          {(weapon.flavourText !== '' || null) && (
            <blockquote>{weapon.flavourText}</blockquote>
          )}

          <p>
            {titleCase(weapon.class)} {weapon.family}
          </p>
          <div className="single-weapon__title-bar">
            <h3 className="single-weapon__subheading">Base Damage</h3>
          </div>
          <p>{humanReadableNumber(weapon.damageLevel40)} (Level 40)</p>
          <p>{humanReadableNumber(weapon.damageWt5) || 'N/A'} (World Tier 5)</p>
          <div className="single-weapon__title-bar">
            <h3 className="single-weapon__subheading">Specifications</h3>
          </div>
          {weapon.maxBonusValue > 1 && (
            <p>
              {weapon.maxBonusValue}% {deAbbreviate(weapon.weaponBonusType)}
            </p>
          )}
          <p>{hsDisplay(weapon.headshotMultiplier)} x Headshot Multiplier</p>
          <p>{weapon.rpm} rounds per minute rate of fire</p>
          <p>{weapon.magazineSize} round magazine</p>
          {weapon.modSlots === 0 && <p>No adjustable modification slots</p>}
          {weapon.modSlots > 0 && <p>{weapon.modSlots} modification slots</p>}
          <p>{displayPercentage(weapon.accuracy)}% accuracy</p>
          <p>{displayPercentage(weapon.stability)}% stability</p>
          <p>{weapon.optimalRange}m optimal range</p>
          <p>{weapon.maxRange}m max range</p>
          <p>
            {millisecondsToSeconds(weapon.reloadSpeed)} seconds reload speed
          </p>
          <p>
            {millisecondsToSeconds(weapon.reloadSpeedFromEmpty)} seconds reload
            speed (empty magazine)
          </p>
          {(!weapon.accuracy || !weapon.stability) && (
            <>
              <small>Note: 0 indicates no value available</small>
            </>
          )}
        </div>
        <div className="single-weapon__details">
          <div className="single-weapon__title-bar">
            <h3 className="single-weapon__subheading">Stats Comparison</h3>
          </div>

          <Radar data={chartData} options={options} />
        </div>
      </div>
    </>
  );
}
