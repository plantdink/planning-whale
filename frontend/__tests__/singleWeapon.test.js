import { render, screen } from '@testing-library/react';
import SingleWeapon from '../components/SingleWeapon';
import { fakeWeapon, fakeAverageWeapon } from '../lib/testUtils';
import {
  humanReadableNumber,
  hsDisplay,
  displayPercentage,
  millisecondsToSeconds,
} from '../lib/displayNumbers';
import { deAbbreviate } from '../lib/abbreviationsDictionary';

const weapon = fakeWeapon();
const avgWeapon = fakeAverageWeapon();

describe('<SingleWeapon />', () => {
  it('renders the data correctly for a single weapon', () => {
    const { container, debug } = render(
      <SingleWeapon weapon={weapon} avgWeapon={avgWeapon} />
    );
    // debug();

    expect(container).toMatchSnapshot();

    const weaponTitle = screen.getByText(weapon.model);
    expect(weaponTitle).toBeInTheDocument();

    const weaponDamageLvl40 = `${humanReadableNumber(
      weapon.damageLevel40
    )} (Level 40)`;
    const testLvl40WeaponDamage = screen.getByText(weaponDamageLvl40);
    expect(testLvl40WeaponDamage).toBeInTheDocument();

    const weaponDamageWT5 = `${humanReadableNumber(
      weapon.damageWt5
    )} (World Tier 5)`;
    const testWT5WeaponDamage = screen.getByText(weaponDamageWT5);
    expect(testWT5WeaponDamage).toBeInTheDocument();

    // flavour text blockquote should not be visible if the weapon has null or an empty string
    if (weapon.flavourText !== null || '') {
      const flavourText = container.querySelector('blockquote');
      expect(flavourText).not.toBeInTheDocument();
    }

    const weaponTestBonusDamage = `${weapon.maxBonusValue}% ${deAbbreviate(
      weapon.weaponBonusType
    )}`;
    // https://polvara.me/posts/five-things-you-didnt-know-about-testing-library
    const screenBonusDamage = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestBonusDamage;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenBonusDamage).toBeInTheDocument();

    const weaponTestHSDMultiplier = `${hsDisplay(
      weapon.headshotMultiplier
    )} x Headshot Multiplier`;
    const screenHSDMultiplier = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestHSDMultiplier;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenHSDMultiplier).toBeInTheDocument();

    const weaponTestRPM = `${weapon.rpm} rounds per minute rate of fire`;
    const screenRPM = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestRPM;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenRPM).toBeInTheDocument();

    const weaponTestMagazineSize = `${weapon.magazineSize} round magazine`;
    const screenMagazineSize = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestMagazineSize;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenMagazineSize).toBeInTheDocument();

    const weaponTestModSlots = `${weapon.modSlots} modification slots`;
    const screenModSlots = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestModSlots;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenModSlots).toBeInTheDocument();

    const weaponTestAccuracy = `${displayPercentage(
      weapon.accuracy
    )}% accuracy`;
    const screenAccuracy = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestAccuracy;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenAccuracy).toBeInTheDocument();

    const weaponTestStability = `${displayPercentage(
      weapon.stability
    )}% stability`;
    const screenStability = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestStability;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenStability).toBeInTheDocument();

    const weaponTestOptimalRange = `${weapon.optimalRange}m optimal range`;
    const screenOptimalRange = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestOptimalRange;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenOptimalRange).toBeInTheDocument();

    const weaponTestMaxRange = `${weapon.maxRange}m max range`;
    const screenMaxRange = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestMaxRange;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenMaxRange).toBeInTheDocument();

    const weaponTestReloadSpeed = `${millisecondsToSeconds(
      weapon.reloadSpeed
    )} seconds reload speed`;
    const screenReloadSpeed = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestReloadSpeed;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenReloadSpeed).toBeInTheDocument();

    const weaponTestEmptyReloadSpeed = `${millisecondsToSeconds(
      weapon.reloadSpeedFromEmpty
    )} seconds reload speed (empty magazine)`;
    const screenEmptyReload = screen.getByText((content, node) => {
      const hasText = (node) => node.textContent === weaponTestEmptyReloadSpeed;
      const nodeHasText = hasText(node);
      const childrenDontHaveText = Array.from(node.children).every(
        (child) => !hasText(child)
      );
      return nodeHasText && childrenDontHaveText;
    });
    expect(screenEmptyReload).toBeInTheDocument();

    if (weapon.accuracy || weapon.stability > 0) {
      const small = container.querySelector('small');
      expect(small).not.toBeInTheDocument();
    }

    // it('Renders a Radar Chart', () => {
    //   const wrapper = mount(
    //     <RadarChart />
    //   )
    //   expect(wrapper).toMatchSnapshot();
  });
});
