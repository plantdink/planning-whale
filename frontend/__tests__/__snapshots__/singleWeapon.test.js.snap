// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleWeapon /> renders the data correctly for a single weapon 1`] = `
<div>
  <div
    class="single-weapon__content"
  >
    <div
      class="single-weapon__details"
    >
      <div
        class="single-weapon__title-bar"
      >
        <h1
          class="single-weapon__heading"
        >
          Test ACR-E
        </h1>
      </div>
      <blockquote />
      <p>
        Assault Rifle
         
        Test ACR
      </p>
      <div
        class="single-weapon__title-bar"
      >
        <h3
          class="single-weapon__subheading"
        >
          Base Damage
        </h3>
      </div>
      <p>
        53,124
         (Level 40)
      </p>
      <p>
        11,770
         (World Tier 5)
      </p>
      <div
        class="single-weapon__title-bar"
      >
        <h3
          class="single-weapon__subheading"
        >
          Specifications
        </h3>
      </div>
      <p>
        21
        % 
        Damage to Health
      </p>
      <p>
        1.55
         x Headshot Multiplier
      </p>
      <p>
        650
         rounds per minute rate of fire
      </p>
      <p>
        30
         round magazine
      </p>
      <p>
        4
         modification slots
      </p>
      <p>
        61.8
        % accuracy
      </p>
      <p>
        76
        % stability
      </p>
      <p>
        27
        m optimal range
      </p>
      <p>
        105
        m max range
      </p>
      <p>
        2.00
         seconds reload speed
      </p>
      <p>
        2.50
         seconds reload speed (empty magazine)
      </p>
    </div>
    <div
      class="single-weapon__details"
    >
      <div
        class="single-weapon__title-bar"
      >
        <h3
          class="single-weapon__subheading"
        >
          Stats Comparison
        </h3>
      </div>
    </div>
  </div>
</div>
`;
